lua/neoclip/db.lua

```diff
@@ -0,0 +1,70 @@
+local M = {}
+
+local has_sqlite, sqlite = pcall(require, "sqlite")
+if not has_sqlite then
+    print "Couldn't find sqlite.lua. Cannot use persistent history"
+    return nil
+end
+
+local function get_db(db_path)
+    local db = sqlite.new(db_path)
+    db:open()
+    db:create('neoclip', {
+        id = true,
+        ensure = true,
+        regtype = 'text',
+        contents = 'text',
+        filetype = 'text',
+    })
+    return db
+end
+
+local function copy(t)
+    local new = {}
+    for k, v in pairs(t) do
+        if type(v) == 'table' then
+            new[k] = copy(v)
+        else
+            new[k] = v
+        end
+    end
+    return new
+end
+
+local function map(t, f)
+    local new_t = copy(t)
+    for k, v in pairs(new_t) do
+        new_t[k] = f(v)
+    end
+    return new_t
+end
+
+local function to_sqlite(storage)
+    local tbl = map(storage, function(v)
+        v.contents = table.concat(v.contents, '\n')
+        return v
+    end)
+    return tbl
+end
+
+local function from_sqlite(tbl)
+    local storage = map(tbl, function(v)
+        v.contents = vim.fn.split(v.contents, '\n')
+        return v
+    end)
+    return storage
+end
```

Comment:
you can insert lua table to sqlite. just set contents to be of type luatable

Reply:
Ah I see! I tried with just `"table"`.

Edit: Actually are other types supported? I get the error:
```
E5108: Error executing lua ...im/site/pack/packer/start/sqlite.lua/lua/sqlite/stmt.lua:309:         sqlite.lua error at stmt:bind(): Unr
ecognized or unsupported type.
        Please report issue.
```
and looking at the relevant check in `sqlite.lua`:
```lua
      if ((type(v) == "string" and v:match "%a+%(.+%)") and flags.ok or self:bind(index, v)) ~= flags.ok then
        error("sqlite.lua error at stmt:bind(), failed to bind a given value '%s'. Please report issue."):format(v)
      end
```
it seems only strings are supported?

Reply:
You need to edit the type of the schema key to luatable instead of text for this to work.

Reply:
Thanks! Yes it works :+1:, I think at first it didn't work since I still had an local sqlite file with the old schema :)

Reply:
> Thanks! Yes it works :+1:, I think at first it didn't work since I still had an local sqlite file with the old schema :)

😭 it should have auto-alter the schema in the file. I need to add a test for that. Thanks